flytekit/clis/auth/auth.py

```diff
@@ -0,0 +1,234 @@
+
+
+class OAuthCallbackHandler(BaseHTTPRequestHandler):
+    """
+    A simple wrapper around BaseHTTPServer.BaseHTTPRequestHandler that handles a callback URL that accepts an
+    authorization token.
+    """
+
+    def do_GET(self):
+        url = urlparse(self.path)
+        if url.path == self.server.redirect_path:
+            self.send_response(StatusCodes.OK)
+            self.end_headers()
+            self.handle_login(dict(parse_qsl(url.query)))
+        else:
+            self.send_response(404)
+
+    def handle_login(self, data):
+        self.server.handle_authorization_code(AuthorizationCode(data['code'], data['state']))
+
+
+class OAuthHTTPServer(HTTPServer):
+    """
+    A simple wrapper around the BaseHTTPServer.HTTPServer implementation that binds an authorization_client for handling
+    authorization code callbacks.
+    """
+    def __init__(self, server_address, RequestHandlerClass, bind_and_activate=True,
+                 redirect_path=None, queue=None):
+        HTTPServer.__init__(self, server_address, RequestHandlerClass, bind_and_activate)
+        self._redirect_path = redirect_path
+        self._auth_code = None
+        self._queue = queue
+
+    @property
+    def redirect_path(self):
+        return self._redirect_path
+
+    def handle_authorization_code(self, auth_code):
+        self._queue.put(auth_code)
+
+
+class Credentials(object):
+    # TODO(katrogan): Also add expires_in handling.
+    def __init__(self, access_token=None, id_token=None):
+        self._access_token = access_token
+        self._id_token = id_token
+
+    @property
+    def access_token(self):
+        return self._access_token
+
+    @property
+    def id_token(self):
+        return self._id_token
+
+
+# TODO:
+#  do we need to support initiate login URI? https://devforum.okta.com/t/initiate-login-uri-for-all-subdomain-urls/3766
+
+
+class AuthorizationClient(object):
+    def __init__(self, auth_endpoint=None, token_endpoint=None, client_id=None, redirect_path=None):
+        self._auth_endpoint = auth_endpoint
+        self._token_endpoint = token_endpoint
+        self._client_id = client_id
+        self._redirect_path = redirect_path
+        self._code_verifier = generate_code_verifier()
+        code_challenge = create_code_challenge(self._code_verifier)
+        self._code_challenge = code_challenge
+        state = generate_state_parameter()
+        self._state = state
+        self._credentials = None
+
+        self._params = {
+            # TODO: need an audience param here?
+            "client_id": client_id,  # This must match the Client ID of the OAuth application.
+            "response_type": "code",  # Indicates the authorization code grant
+            "scope": "openid",  # ensures that the /token endpoint returns an ID token
+            # callback location where the user-agent will be directed to.
+            "redirect_uri": urljoin("http://localhost:8088", self._redirect_path),
```

Comment:
can we pick a random port that's unused?  random googling, didn't know about this before and haven't tested. https://unix.stackexchange.com/questions/55913/whats-the-easiest-way-to-find-an-unused-local-port

of course this has to be done first, not sure if it is.  https://support.okta.com/help/s/question/0D50Z00008G7VBgSAN/openid-connect-redirect-uri